### ISSUE TITLE
5 yolo layers + LSTM?

### ISSUE THREAD

**Issue by WadRex** [AlexeyAB/darknet#4281]
Hello,
Is there a variation of yolov3_5l.cfg with LSTM to use it on video?
Thanks in advance!

**Comment by AlexeyAB** [comment]
Currently no. 


You can try to implement it by yourself. Or you can try to re-route layer after [upsample] to the eralier layers like there: 

https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> for training for small objects (smaller than 16x16 after the image is resized to 416x416) - set layers = -1, 11 instead of https://github.com/AlexeyAB/darknet/blob/6390a5a2ab61a0bdf6f1a9a6b4a739c16b36e0d7/cfg/yolov3.cfg#L720 and set stride=4 instead of https://github.com/AlexeyAB/darknet/blob/6390a5a2ab61a0bdf6f1a9a6b4a739c16b36e0d7/cfg/yolov3.cfg#L717

**Comment by WadRex** [comment]
> You can try to implement it by yourself.
I am gonna try that tomorrow. In case of success, I'll post cfg in this thread.

**Comment by WadRex** [comment]
@AlexeyAB 

Hello,
I started testing out yolov3_5l on your dataset from that post https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-494148968 and I am getting around 40% accuracy after 3000 iterations, which is worse than tiny yolo with 3 layers and I dont believe it's gonna go up, since it kinda converged.
Did you trained other models with pre-trained weights? If yes, will those "http://pjreddie.com/media/files/darknet53.conv.74" be suitable for 5 layers model?

**Comment by WadRex** [comment]
@AlexeyAB Were the pretrained weights used in https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-494148968
If yes, will those "darknet53.conv.74" be usable?

**Comment by AlexeyAB** [comment]
@WadRex Yes: https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-494154586

for Full yolov3 use http://pjreddie.com/media/files/darknet53.conv.74

**Comment by nyj-ocean** [comment]
@AlexeyAB 
In` yolo_v3_spp_lstm.cfg` , I found `[conv_lstm]` was set in 3 `yolo detect layers` , not in the `darknet53` 

So in `yolov3_5l.cfg` , could I set  `[conv_lstm]` in 5 `yolo detect layers`  to get `Lstm+yolov3_5l.cfg`, and then use `darknet53.conv.74`  to train `Lstm+yolov3_5l.cfg` ?

Is it possible?

**Comment by AlexeyAB** [comment]
@nyj-ocean Yes.

You can add as many conv-lstm-layers as you like while you have enough GPU-memory to hold `timesteps=2` or more

**Comment by WadRex** [comment]
@AlexeyAB 
Well, it's been a long time.
I trained yolov3_5l on your dataset from this issue https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-494148968
The problem is, I was messing with my own datasets and because of that I had uploaded the wrong .cfg file. (I am training on a server). The batch size is "5" instead of "64". I noticed it very late into the training process, so I let it run to the end.
But I wouldn't be writing this comment if the results were bad as I was expecting them to be.
Not looking and batch size, the results were better then all of the existed models in this issue: https://github.com/AlexeyAB/darknet/issues/3114#issuecomment-494148968-
The best map was 76.16% and I would expect it go be even higher with bigger batch size.
This is without LSTM aswell, so higher batch size and LSTM in 5 layers model should give a very good result.
I attached the chart and if you need, I can attach best and last weights aswell.
![chart](https://user-images.githubusercontent.com/47948892/70866015-88c4a200-1f64-11ea-9c68-097dab0b25df.png)

**Comment by AlexeyAB** [comment]
@WadRex Did you train this cfg-file https://github.com/AlexeyAB/darknet/blob/master/cfg/yolov3_5l.cfg with batch=5 subdivisions=5 with pre-trained weights `darknet53.conv.74` ?

Yes, I know that this model is very accurate especially for small objects.
Try to train `yolov3_5l_lstm.cfg` with batch=64 subdivisions=32 or 64 and `time_steps >= 2`

**Comment by WadRex** [comment]
@AlexeyAB 
> 
> @WadRex Did you train this cfg-file https://github.com/AlexeyAB/darknet/blob/master/cfg/yolov3_5l.cfg with batch=5 subdivisions=5 with pre-trained weights `darknet53.conv.74` ?
> 
I trained with batch=5 and subdivisions=1 with those pre-trained weights.

I am going to train LSTM version later this week, will let you know the results.
